### ISSUE TITLE
[xpu-alignment] _adaptive_avg_pool2d_backward on XPU accepts a leading-dim mismatch between input and grad_output

### ISSUE THREAD

**Issue by laifenxiawucha** [intel/torch-xpu-ops#5233]
## Independent local verification

XPU rerun 2026-09-07 (Max 1100, torch 2.15.0.dev20260902+xpu): direct aten._adaptive_avg_pool2d_backward accepted both channel-mismatch directions and returned input-shaped gradients instead of raising.

The upstream result is still tracked as upstream-dependent; this issue records the concrete XPU defect and does not assert that upstream has finalized cross-backend semantics.

<sub>alignment scan `2026-08-30`, run `33350143612`</sub>
## Summary

Upstream source: pytorch/pytorch#195339 (open) - `Validate leading dims of input and grad_output in _adaptive_avg_pool2d_backward`, which fixes pytorch/pytorch#194803 (`[CPU] _adaptive_avg_pool2d_backward accesses out of bounds for mismatched channels`, labelled `module: crash` and `security`).
Alignment scan window: `[2026-08-30T00:00:00Z, 2026-08-31T00:00:00Z)` over `pytorch/pytorch`.

This is the 2D counterpart of pytorch/pytorch#195304 in a distinct operator and a distinct XPU file, not a second reproduction of the same code path. `grad_input` is sized from `input` while the kernel takes its channel bound from whichever tensor its memory-format branch prefers, so a leading-dim mismatch strides one pointer past the end of the other tensor. The upstream PR adds the per-dimension check to the CPU helper and, separately, to the CUDA template; XPU has its own kernel and did not receive it.

## Observed XPU behavior

Every mismatch is accepted and the call returns an input-shaped tensor, the pre-fix signature; the tested build predates the fix, which is why CPU behaves the same way:

```
contiguous, grad_output has more channels: input=(1, 1, 1, 1) grad_output=(1, 2, 1, 1)
  cpu: returned (1, 1, 1, 1)
  xpu: returned (1, 1, 1, 1)
channels_last, grad_output has fewer channels: input=(1, 2, 1, 1) grad_output=(1, 1, 1, 1)
  cpu: returned (1, 2, 1, 1)
  xpu: returned (1, 2, 1, 1)
contiguous, input has more channels: input=(1, 2, 1, 1) grad_output=(1, 1, 1, 1)
  cpu: returned (1, 2, 1, 1)
  xpu: returned (1, 2, 1, 1)
```

After the upstream fix CPU and CUDA raise `adaptive_avg_pool2d_backward: input and grad_output must have the same size at dim <i>, but got <a> and <b>`.

Scope of the evidence: the runs establish that the guard is absent and that the call returns instead of raising; they do not demonstrate the out-of-bounds read itself, which needs an ASan build. The second case is labelled `channels_last`, but a `(1, 2, 1, 1)` tensor is contiguous in both formats, so the channels-last branch is not separately proven. The two contiguous cases carry the finding on their own, and the missing check is visible in the source for both branches.

## Reproducer

```python
import torch

CASES = (
    ("grad_output has more channels", (1, 1, 1, 1), (1, 2, 1, 1)),
    ("input has more channels", (1, 2, 1, 1), (1, 1, 1, 1)),
)

for label, input_shape, grad_shape in CASES:
    torch.manual_seed(0)
    x = torch.randn(input_shape, device="xpu")
    grad_output = torch.randn(grad_shape, device="xpu")
    try:
        out = torch.ops.aten._adaptive_avg_pool2d_backward(grad_output, x)
        torch.xpu.synchronize()
        print(label, "returned", tuple(out.shape))
    except RuntimeError as exc:
        print(label, "raised", str(exc).splitlines()[0])
```

## Target path

`_adaptive_avg_pool2d_backward` dispatches per backend (`CPU`, `CUDA`, `MPS`, `XPU: adaptive_avg_pool2d_backward_xpu`), so there is no shared validation layer.

- `src/ATen/native/xpu/AdaptiveAveragePooling2d.cpp`: `adaptive_avg_pool2d_backward_xpu` performs `adaptive_pool_empty_output_check`, the rank equality check and the 3D/4D check, and nothing that compares leading dims.
- `src/ATen/native/xpu/sycl/AdaptiveAveragePooling2dKernels.cpp`: `adaptive_avg_pool2d_backward_kernel` selects its memory-format branch from `input_.suggest_memory_format()` and takes `sizeC` from `input.size(-3)` and `sizeB` from `input.size(0)` while indexing `grad_output` with `grad_output`'s own strides, so the two tensors are walked with independent channel counts.

The fix mirrors the upstream hunk: for `i` in `[0, ndim - 2)`, `TORCH_CHECK(input.size(i) == grad_output.size(i), ...)` with the upstream message text, alongside the existing rank check in the dispatch wrapper.

## Ownership

The upstream PR touches only `aten/src/ATen/native/AdaptiveAveragePooling.cpp` and `aten/src/ATen/native/cuda/AdaptiveAveragePooling.cu` plus a test; neither file is on the XPU dispatch path. No upstream issue or PR claims the XPU side, and no existing `intel/torch-xpu-ops` issue or PR covers this validation. Independent XPU work is required here.

## Environment

```
torch 2.15.0.dev20260830+xpu
python 3.13.13 (/opt/conda/bin/python3.13)
torch path /opt/conda/lib/python3.13/site-packages/torch/__init__.py
xpu device Intel(R) Arc(TM) Pro B60 Graphics
```

Filed by the automated XPU alignment pipeline after independent review of the scan evidence. The reproducer above was executed by the deterministic runner, not by the reviewing agent.

**Comment by laifenxiawucha** [comment]
@torchxpubot review

**Comment by torchxpubot** [comment]
`@torchxpubot review` can only be used on pull request comments.

**Comment by laifenxiawucha** [comment]
@torchxpubot fix

**Comment by torchxpubot** [comment]
Automated fix session for issue #5233.



## Issue Triage

| Field | Value |
|-------|-------|
| Issue type | single-bug |
| Reproduction missing | no |
| Scope | torch-xpu-ops |
| Dependencies | (none) |
| Handling | agent-fixable |

## Reproduce

| Field | Value |
|-------|-------|
| Verdict | **REPRODUCED** |
| Stage | nightly wheel |
| Wheel | `torch 2.15.0.dev20260905+xpu` |
| Device | Intel(R) Arc(TM) Pro B60 Graphics |

All six leading-dim mismatch shapes (4D contiguous both directions, 4D
channels-last, 4D batch mismatch, 3D channel mismatch) are accepted by
`aten::_adaptive_avg_pool2d_backward` on XPU and return an `input`-shaped
tensor instead of raising.

Runner note: the GPUs on this host are currently wedged (`xe` GuC job
timeouts on every device, reproducible with a plain SYCL `p[i] = i + 1`
kernel), so the reproducer deliberately allocates with `torch.empty` and
never synchronizes. The missing check is host-side, so this still exercises
exactly the validation being fixed without depending on kernel execution.



## Root Cause

| Field | Value |
|-------|-------|
| Verdict | **IMPLEMENTING** (ok) |
| target_repo | `torch-xpu-ops` |
| domains | `["xpu-kernel"]` |
| analyzed_sha | `180f7f9296e724fc5b80ba4c273c3cad4e501d7e` |

`adaptive_avg_pool2d_backward_xpu`
(`src/ATen/native/xpu/AdaptiveAveragePooling2d.cpp`) validates only that
`input` and `grad_output` have the *same rank* and that the rank is 3 or 4.
It never compares the leading (batch/channel) extents, so a mismatched
`grad_output` reaches the kernel.

Downstream, the extents are taken from the wrong tensor:

- In `adaptive_avg_pool2d_backward_kernel`
  (`src/ATen/native/xpu/sycl/AdaptiveAveragePooling2dKernels.cpp`), `sizeB`
  and `sizeC` come from `input`, while `ostrideB/C/H/W` come from
  `grad_output`. For the 3D path it also reshapes `grad_output` to
  `{1, sizeC, osizeH, osizeW}` using `input`'s channel count.
- `AdaptiveAvgPool2dBwdSLMKernelFunctor` derives its batch/channel bounds
  from `gxacc` (i.e. `grad_input`, sized from `input`) and indexes
  `gyacc_[_ob][_oc][_oh][_ow]` with those same indices.

`PackedTensorAccessor64` does no bounds checking, so an oversized index into
`grad_output` is a plain out-of-bounds read.

Upstream PyTorch PR
[pytorch/pytorch#195339](https://github.com/pytorch/pytorch/pull/195339)
adds exactly this guard, but only to the CPU
(`aten/src/ATen/native/AdaptiveAveragePooling.cpp`) and CUDA
(`aten/src/ATen/native/cuda/AdaptiveAveragePooling.cu`) implementations;
neither file is on the XPU dispatch path, so XPU needs its own copy. That PR
is still open, so its extended `test_adaptive_avg_pooling_backward_fails`
cases are not yet inherited by `test/xpu/nn/test_pooling_xpu.py`.

## Proposed Fix Strategy

Add the upstream leading-dim check to the XPU dispatch wrapper, mirroring
upstream's condition and message text so the eventual shared upstream test
passes unchanged on XPU. Land a local regression test under
`test/regressions/` in the meantime.



## Implement

Branch `agent/fix-issue-5233`, one commit on top of `180f7f9`.

```diff
diff --git a/src/ATen/native/xpu/AdaptiveAveragePooling2d.cpp b/src/ATen/native/xpu/AdaptiveAveragePooling2d.cpp
index be3c86fe..9c68bae0 100644
--- a/src/ATen/native/xpu/AdaptiveAveragePooling2d.cpp
+++ b/src/ATen/native/xpu/AdaptiveAveragePooling2d.cpp
@@ -11,6 +11,7 @@
 #include <ATen/WrapDimUtilsMulti.h>
 #include <ATen/core/Tensor.h>
 #include <ATen/native/AdaptivePooling.h>
+#include <c10/util/irange.h>
 
 #include <comm/xpu_aten.h>
 
@@ -45,6 +46,21 @@ Tensor adaptive_avg_pool2d_backward_xpu(
       (input.ndimension() == 3 || input.ndimension() == 4),
       "non-empty 3D or 4D (batch mode) tensor expected for input");
 
+  // The kernel indexes grad_output with the batch/channel extents taken from
+  // input, so a leading-dim mismatch reads out of bounds (grad_output smaller)
+  // or silently ignores part of it (grad_output larger).
+  for (const auto i : c10::irange(input.dim() - 2)) {
+    TORCH_CHECK(
+        input.size(i) == grad_output.size(i),
+        __func__,
+        ": input and grad_output must have the same size at dim ",
+        i,
+        ", but got ",
+        input.size(i),
+        " and ",
+        grad_output.size(i));
+  }
+
   globalContext().alertNotDeterministic("adaptive_avg_pool2d_backward_xpu");
 
   Tensor grad_input;
diff --git a/test/regressions/test_adaptive_avg_pool2d_bwd.py b/test/regressions/test_adaptive_avg_pool2d_bwd.py
new file mode 100644
index 00000000..4563098f
--- /dev/null
+++ b/test/regressions/test_adaptive_avg_pool2d_bwd.py
@@ -0,0 +1,60 @@
+# Copyright 2020-2026 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+
+# Owner(s): ["module: intel"]
+import torch
+from torch.testing._internal.common_utils import (
+    instantiate_parametrized_tests,
+    parametrize,
+    run_tests,
+    subtest,
+    TestCase,
+)
+
+aten = torch.ops.aten
+
+
+class TestAdaptiveAvgPool2dBwd(TestCase):
+    # Regression test: the backward kernel indexes grad_output with the
+    # batch/channel extents taken from input, so a leading-dim mismatch used to
+    # read out of bounds instead of raising. Allocation-only (torch.empty, no
+    # synchronize) because the check under test is host-side.
+    @parametrize(
+        "input_shape, grad_output_shape, memory_format",
+        [
+            subtest(
+                ((1, 2, 3, 3), (1, 3, 7, 7), torch.contiguous_format),
+                name="channel_4d",
+            ),
+            subtest(
+                ((1, 3, 3, 3), (1, 2, 7, 7), torch.channels_last),
+                name="channel_4d_channels_last",
+            ),
+            subtest(
+                ((2, 2, 3, 3), (1, 2, 7, 7), torch.contiguous_format),
+                name="batch_4d",
+            ),
+            subtest(((2, 3, 3), (3, 7, 7), torch.contiguous_format), name="channel_3d"),
+        ],
+    )
+    def test_leading_dim_mismatch_raises(
+        self, input_shape, grad_output_shape, memory_format
+    ):
+        x = torch.empty(input_shape, device="xpu", memory_format=memory_format)
+        grad_output = torch.empty(
+            grad_output_shape, device="xpu", memory_format=memory_format
+        )
+        with self.assertRaisesRegex(RuntimeError, "must have the same size at dim"):
+            aten._adaptive_avg_pool2d_backward(grad_output, x)
+
+
+instantiate_parametrized_tests(TestAdaptiveAvgPool2dBwd)
+
+
+if __name__ == "__main__":
+    run_tests()
```



## Verify

| Field | Value |
|-------|-------|
| Verdict | **PASSED** |
| Stage | source build |
| Build | `torch 2.15.0a0+git990e00e` (pytorch `990e00eb`, `TORCH_XPU_ARCH_LIST=bmg,pvc`) |
| torch-xpu-ops | `agent/fix-issue-5233` @ `dc06782`, pinned via `third_party/xpu.txt` |

The new check is confirmed present in the built `libtorch_xpu.so`, and `torch`
imports from the source tree, so the before/after signal comes from this diff.

- **Reproducer** `agent_space_xpu/repro_5233.py`: was `FAILED: 6/6 cases not
  guarded` on the nightly wheel, now `PASSED: all 6 cases raised the
  leading-dim check` (exit 0). Every case reports the expected dim index and
  the two sizes.
- **Regression test** `pytest -sv test/regressions/test_adaptive_avg_pool2d_bwd.py`:
  4 passed.
- **Negative control**: a matching-leading-dim call
  (`input (2,3,0,0)`, `grad_output (2,3,7,7)`, which takes the `numel()==0`
  `zeros_like` path and launches no kernel) is still accepted, so the guard
  does not over-reject.
- **Lint**: `lintrunner -a -m origin/main` and `lintrunner --take CLANGTIDY`
  (with `PYTORCH_ROOT` set) both report no issues; nothing to amend.

Caveat: the GPUs on this runner are wedged (every device hangs or returns
`UR_RESULT_ERROR_OUT_OF_RESOURCES` on a plain `a @ a`, independently of this
issue), so no kernel-executing test could be run here. Verification is
therefore limited to the host-side validation path, which is exactly what this
diff changes; the numerics of the valid backward path are untouched and remain
covered by `test/xpu/nn/test_pooling_xpu.py` in normal CI.



## Summary

**Verdict: DONE** — fix verified, waiting on a human to open the PR.

| Field | Value |
|-------|-------|
| Branch | `agent/fix-issue-5233` |
| Base | `180f7f9296e724fc5b80ba4c273c3cad4e501d7e` |
| Commit | `dc06782` (1 commit) |
| Files | `src/ATen/native/xpu/AdaptiveAveragePooling2d.cpp`, `test/regressions/test_adaptive_avg_pool2d_bwd.py` |

`adaptive_avg_pool2d_backward_xpu` checked rank but not the leading dims, while
the kernel indexes `grad_output` using batch/channel extents taken from
`input` -- an unchecked out-of-bounds read via `PackedTensorAccessor64`. The fix
adds the missing check to the dispatch wrapper, using upstream's condition and
message text verbatim so the extended
`TestPoolingNNDevice.test_adaptive_avg_pooling_backward_fails` cases from
[pytorch/pytorch#195339](https://github.com/pytorch/pytorch/pull/195339) will
pass on XPU unchanged once that PR lands. A local regression test covers it
until then.

Nothing was pushed and no PR was opened; the branch is local for a human to
review and submit.

### Caveats and follow-ups

- The GPUs on this runner are wedged for compute (all 8 devices hang or fail a
  plain matmul, independently of this issue), so verification covered the
  host-side validation path only -- which is all this diff touches. The valid
  backward path is unchanged.
- `adaptive_avg_pool3d_backward` on XPU has the identical defect
  (`src/ATen/native/xpu/AdaptiveAveragePooling3d.cpp` and its kernels take
  `sizeD`/`sizeB` from `input` while striding `gradOutput`). Upstream #195339
  does not cover 3D either. Left out to keep this change scoped to the reported
  issue -- worth a follow-up.
- The terminal `agent:done` label does not exist in this repo (only
  `agent:active`, `agent:needs-human`, `agent:blocked`, `agent:skipped`,
  `agent:close`, `agent:merged`), so no terminal label was set rather than
  picking a misleading one. The issue carries no stale `agent:active`.
- `meta__adaptive_avg_pool2d_backward` in PyTorch has no equivalent check, so
  the guard does not fire under fake-tensor/meta tracing. That is an upstream
  gap, not something this repo can fix.

*Automated by issue-handler.*

_Generated by [fix job](https://github.com/intel/torch-xpu-ops/actions/runs/34106137117)._

**Comment by laifenxiawucha** [comment]
Closing as premature. Verified on an Intel Data Center GPU Max 1100 (`torch 2.12.1+xpu`): CPU accepts the same mismatched leading dims without raising, so this is a validation gap shared by all backends, not a CPU-vs-XPU divergence. The out-of-bounds read is real at the source level (the kernel takes its batch/channel extents from `input` while walking `grad_output` with its own strides) but was never directly demonstrated; an ASan-instrumented build would be needed.

The upstream contract is still unsettled: pytorch/pytorch#195339 has no review. Landing an XPU-only `TORCH_CHECK` now would make XPU the only backend rejecting these shapes, with error text that may not match what upstream settles on.

**This issue is not a canonical tracker.** If #195339 merges, the merge event re-enters the alignment candidate set and a fresh issue should be filed. Closed together with #5232, which is the same gap in the 3D kernel.

**Comment by laifenxiawucha** [comment]
**weak-rejected** 

why: filter-wrong
notes: Same defect as #5232 in the 2D kernel and the same reason to wait
for pytorch/pytorch#195339; the patch additionally puts the guard in the
dispatch wrapper, which an in-repo caller reaching the kernel directly
bypasses, and its test uses `torch.empty` with no valid-shape control.
